glyphs-reader/src/propagate_anchors.rs

```diff
@@ -0,0 +1,883 @@
+        a.pos -= origin;
+        (a.name.clone(), a)
+    }));
+    let mut has_underscore_anchor = false;
+    let mut has_mark_anchor = false;
+    let mut component_count_from_anchors = 0;
+
+    // now we count how many components we have, based on our anchors
+    for name in all_anchors.keys() {
+        has_underscore_anchor |= name.starts_with('_');
+        has_mark_anchor |= name.chars().next().unwrap_or('\0').is_ascii_alphabetic();
+        if !is_ligature
+            && number_of_base_glyphs == 0
+            && !name.starts_with('_')
+            && !(name.ends_with("entry") | name.ends_with("exit"))
+            && name.contains('_')
+        {
+            let (_, suffix) = name.split_once('_').unwrap();
+            // carets count space between components, so the last caret
+            // is n_components - 1
+            let maybe_add_one = name.starts_with("caret").then_some(1).unwrap_or_default();
+            let anchor_index = suffix.parse::<usize>().unwrap_or(0) + maybe_add_one;
+            component_count_from_anchors = component_count_from_anchors.max(anchor_index);
+        }
+    }
+    if !has_underscore_anchor && number_of_base_glyphs == 0 && has_mark_anchor {
+        number_of_base_glyphs += 1;
+    }
+    if layer.anchors.iter().any(|a| a.name == "_bottom") {
+        all_anchors.shift_remove("top");
+        all_anchors.shift_remove("_top");
+    }
+    if layer.anchors.iter().any(|a| a.name == "_top") {
+        all_anchors.shift_remove("bottom");
+        all_anchors.shift_remove("_bottom");
+    }
+    base_glyph_counts.insert(
+        (glyph.name.clone(), layer.layer_id.as_str()),
+        number_of_base_glyphs,
+    );
+    all_anchors.into_values().collect()
+}
+
+// returns a vec2 where for each axis a negative value indicates that axis is considerd flipped
+fn get_xy_rotation(xform: Affine) -> Vec2 {
+    // this is based on examining the behaviour of glyphs via the macro panel
+    // and careful testing.
+    let [xx, xy, ..] = xform.as_coeffs();
+    // first take the rotation
+    let angle = xy.atan2(xx);
+    // then remove the rotation, and take the scale
+    let rotated = xform.pre_rotate(-angle).as_coeffs();
+    let mut scale = Vec2::new(rotated[0], rotated[3]);
+    // then invert the scale if the rotation was >= 180°
+    if (angle.to_degrees() - 180.0).abs() < 0.001 {
+        scale *= -1.0;
+    }
+
+    scale
+}
+
+// apply the transform but also do some rounding, so we don't have anchors
+// with points like (512, 302.000000006)
+fn apply_transform_to_anchor(anchor: &mut Anchor, transform: Affine) {
+    // how many zeros do we care about? not this many
+    const ROUND_TO: f64 = 1e6;
+    let mut pos = (transform * anchor.pos).to_vec2();
+    pos *= ROUND_TO;
+    pos = pos.round();
+    pos /= ROUND_TO;
+    anchor.pos = pos.to_point();
+}
+
+fn maybe_rename_component_anchor(comp_name: SmolStr, anchors: &mut [Anchor]) {
+    // e.g, go from 'top_1' to 'top'
```

Comment:
Just to confirm, we definitely want the _first_ `_` or assume there is only one of them?

Reply:
the original code is assuming only one.